lib/bolt/config/transport.rb

```diff
@@ -0,0 +1,133 @@
+        @boltdir = boltdir
+        validate
+      end
+
+      def [](key)
+        @config[key]
+      end
+
+      def to_h
+        @config
+      end
+
+      def fetch(*args)
+        @config.fetch(*args)
+      end
+
+      def include?(args)
+        @config.include?(args)
+      end
+
+      def config=(data)
+        assert_hash(data)
+        @config = data
+        validate
+      end
+
+      def merge(*data)
+        merge_data = {}
+
+        data.compact.each do |d|
+          assert_hash(d)
+          merge_data = Bolt::Util.deep_merge(merge_data, d)
+        end
+
+        merged = Bolt::Util.deep_merge(@config, merge_data)
+        self.class.new(merged, @boltdir)
+      end
+
+      def self.options
+        unless defined? self::OPTIONS
+          raise NotImplementedError,
+                "Constant OPTIONS must be implemented by the transport config class"
+        end
+        self::OPTIONS.keys
+      end
+
+      private def defaults
+        unless defined? self.class::DEFAULTS
+          raise NotImplementedError,
+                "Constant DEFAULTS must be implemented by the transport config class"
+        end
+        self.class::DEFAULTS
+      end
+
+      private def filter(unfiltered)
+        unfiltered.slice(*self.class.options)
+      end
+
+      private def assert_hash(data)
+        return if data.is_a?(Hash)
+        raise Bolt::ValidationError,
+              "Transport config must be a Hash, received #{data.class} #{data.inspect}"
+      end
+
+      private def normalize_interpreters(interpreters)
+        Bolt::Util.walk_keys(interpreters) do |key|
+          key.chars[0] == '.' ? key : '.' + key
+        end
+      end
+
+      # Checks whether a config option contains a plugin reference, which
+      # should always be a valid input
+      private def reference?(value)
+        value.is_a?(Hash) && value.key?('_plugin')
+      end
+
+      # Generic type validation for config options
+      private def validate_type(type, *opts)
+        opts.each do |opt|
+          next if config[opt].nil?
+
+          value = config[opt]
+          unless value.is_a?(type) || reference?(value)
+            raise Bolt::ValidationError,
+                  "#{opt} must be a #{type}, received #{value.class} #{value.inspect}"
+          end
+        end
+      end
+
+      # Boolean type validation for config options
+      # Ruby doesn't have a Boolean class, so these need to be handled separately
+      # from the generic type validation
+      private def assert_boolean(*opts)
+        opts.each do |opt|
+          next if config[opt].nil?
+
+          value = config[opt]
+          unless !!value == value || reference?(value)
+            raise Bolt::ValidationError,
+                  "#{opt} must be a Boolean true or false, received #{value.class} #{value.inspect}"
+          end
+        end
+      end
+
+      # Default validation? No validation!
+      private def validate; end
```

Comment:
Hmmm what do we do about the case where the _entire_ value is a reference? We could add some logic around that to `validate` but then the transport-specific validation would still run... Currently all of those should work since they check whether the keys they're validating are present first, but some different validation in the future could fail. There's also the fact that we currently will merge the default config values _into_ the reference, which is bad.

Maybe we should just check in the `initialize` method whether the whole value is a reference and skip validating in that case?

There's a similar issue for transport-specific validation where individual values could be references and the transports don't check for that currently. I'm a bit unclear how to handle that though...

Reply:
Updated to skip validation if the `input` to the config object is a reference. It still merges with the defaults, but since the reference will be resolved before any config options are accessed, that should be fine. Once the reference is resolved in `Bolt::Inventory` it will be remerged with the defaults and validated.

Reply:
There still remains the issue that some parameters can't be specified as plugin references because they will either be flagged as invalid or just cause an error. Since `assert_type` skips validating options whose values are references, the subsequent validation in each transport config class can be confronted with potentially invalid data.

For instance:

```yaml
ssh:
  private-key:
    _plugin: yaml
    filepath: key_data.yaml
```

This fails because it expects the value of `private-key` to be a hash containing a `key-data` key, which it isn't... yet.

This will affect:
* SSH: `private-key`, `interpreters` and `run-as-command`
* Docker: `interpreters`
* Local: `interpreters` and `run-as-command`
* Orch: `cacert` and `token-file`
* WinRM: `cacert` and `interpreters`

Admittedly, none of those seem like options you would especially _want_ to use a plugin for (except maybe `private-key`, but just set `key-data` instead of `private-key` and it will work), but it still ought to be accepted.

Given the explicit behavior of `super` here is to skip validation of anything that is a reference, I think it's not unreasonable to require the child class double-check the types of anything it wants to apply its own validation to.

Reply:
Updated to skip validation for each of these options if the value is a reference. Once the plugin references are resolved and the input is overwritten the transport config will be revalidated in full.